File: backend/shared/auth/permissions.go
package auth

import "smartdorm/shared/jwt"

// Permission represents a fine-grained access right
type Permission string

const (
	// Property permissions
	PropertyCreate Permission = "property:create"
	PropertyRead   Permission = "property:read"
	PropertyUpdate Permission = "property:update"
	PropertyDelete Permission = "property:delete"

	// Room permissions
	RoomCreate Permission = "room:create"
	RoomRead   Permission = "room:read"
	RoomUpdate Permission = "room:update"
	RoomDelete Permission = "room:delete"

	// Contract permissions
	ContractView   Permission = "contract:view"
	ContractApply  Permission = "contract:apply"
	ContractSign   Permission = "contract:sign"
	ContractManage Permission = "contract:manage"

	// Invoice/Payment permissions
	InvoiceView Permission = "invoice:view"
	InvoicePay  Permission = "invoice:pay"
	InvoiceEdit Permission = "invoice:edit"
)

// RolePermissions maps a Role to its allowed permissions
var RolePermissions = map[string][]Permission{
	jwt.RoleLandlord: {
		PropertyCreate, PropertyRead, PropertyUpdate, PropertyDelete,
		RoomCreate, RoomRead, RoomUpdate, RoomDelete,
		ContractView, ContractManage, ContractSign,
		InvoiceView, InvoiceEdit,
	},
	jwt.RoleTenant: {
		PropertyRead,
		RoomRead,
		ContractView, ContractApply, ContractSign,
		InvoiceView, InvoicePay,
	},
	jwt.RoleAdmin: {
		// Super admins have all permissions (or special handling)
	},
}

// HasPermission check if a role has a specific permission
func HasPermission(role string, perm Permission) bool {
	perms, ok := RolePermissions[role]
	if !ok {
		// Secure Fallback: Deny if role is undefined
		return false
	}
	for _, p := range perms {
		if p == perm {
			return true
		}
	}
	return false
}

File: frontend/src/shared/lib/auth/permissions.ts
import { jwtDecode } from 'jwt-decode';

/**
 * Permission check utility for UI elements.
 * NOTE: This is only for UI/UX (hiding elements). 
 * Security is enforced strictly on the Backend.
 */

interface Claims {
  active_role: string;
  workspace_id?: string;
}

// Map from Backend (must be kept in sync or fetched from API)
const RolePermissions: Record<string, string[]> = {
  LANDLORD: [
    'property:create', 'property:read', 'property:update', 'property:delete',
    'room:create', 'room:read', 'room:update', 'room:delete',
    'contract:view', 'contract:manage', 'contract:sign',
    'invoice:view', 'invoice:edit'
  ],
  TENANT: [
    'property:read', 'room:read',
    'contract:view', 'contract:apply', 'contract:sign',
    'invoice:view', 'invoice:pay'
  ]
};

export const can = (permission: string): boolean => {
  const token = localStorage.getItem('access_token');
  if (!token) return false;

  try {
    const claims = jwtDecode<Claims>(token);
    const role = claims.active_role;
    const permissions = RolePermissions[role] || [];
    
    return permissions.includes(permission);
  } catch (e) {
    return false;
  }
};
